Exercise 13

CameronB9 opened the pull request:
Image of logs in loggly:
![image](https://github.com/CameronB9/DevOps-Course-Starter/assets/135842865/555881a2-7740-476c-a53b-36d0a2e5c8dd)

JSON logs in loggly:
![image](https://github.com/CameronB9/DevOps-Course-Starter/assets/135842865/800c095c-d303-4748-a50a-2af23bef1776)

Error alert:
![Untitled](https://github.com/CameronB9/DevOps-Course-Starter/assets/135842865/12e34c1d-e782-41a5-9ee5-e4eeccef49fd)

## Pipeline Adjustment
I have changed when the terraformed E2E test runs. It should now only run on a pull request to the main branch. For anything else, the regular E2E test should run. I was running into issues with runs in parallel, one would fail due to the terraform state lock. If I have an open pull request, both the push and pull request trigger a pipeline run.

Commit:
Fix issue with running e2e test locally

Commit:
Updated e2e test

Commit:
Update CI Pipeline - only run e2e tests with test terraform environment on pull request or push to main. Don't run terraform apply when pipeline triggered by daily schedule

Commit:
Added logs for actions such as adding/updating todos.
Send logs to loggly.

Commit:
Update ci pipeline - set log level in unit test step.

Update terraform config. Add log level to app service environment variables.

Commit:
Added monitoring with Prometheus.
Send logs to loggly in JSON format.

Commit:
Send HTTP logs to loggly.
Add loggly token to terraform modules to send production logs to loggly.
Update CI Pipeline to pass in loggly token to terraform prod command.

Commit:
Update README.md

Commit:
Update CI Pipeline - only run terraformed test env on pull request to main. Add a step that runs local e2e tests for anything that isn't pull request to main.

Commit:
CI Pipeline - remove env IS_PULL_REQUEST_MAIN.

Commit:
CI Pipeline - set log level to debug for e2e tests local

Commit:
CI Pipeline - add mongo db name and connection string as env vars to local e2e test step

JackMeadDev commented on .github/workflows/ci-pipeline.yml:
```diff
@@ -27,12 +27,16 @@ jobs:
         name: E2E test app terraform init
         working-directory: terraform/modules/test
         run: terraform init
+        if: github.event.schedule != '08 13 * * *' && github.event_name == 'pull_request' && github.base_ref == 'main'
```
Not one I'm certain on but given that you check the event name is `pull_request`, do we _also_ need to check that it's not the schedule?

Regardless, sensible approach to prevent the TF locking issue 👍

JackMeadDev commented on todo_app/app.py:
```diff
@@ -115,9 +137,33 @@ def login_callback():
         user_management = UserManagement()
         user = user_management.get_user(user_id)
 
+        app.logger.info({
```
Is there a particular reason you've switched between the two formats below?
```python
app.logger.info('message here', extra={'category': Category})
# vs
app.logger.info({'message': 'message here', 'category': Category})
```

CameronB9 replied:
I started with the first format and then discovered the second one. I thought I had changed them all to be the second format, I must have missed some.

JackMeadDev commented on todo_app/user.py:
```diff
@@ -30,7 +41,7 @@ def role(self):
 
     @staticmethod
-    def check_permission(level):
+    def check_permission(level, action: Actions, app: Flask = None):
```
We can access the `app` variable with `current_app` imported from Flask, which should correctly find the app without needing to pass it about directly. Either of these approaches do drive us towards having lots of functions which are inherently dependent on Flask though, so it might be that we provide an internal function for `get_logger` or similar that wraps that at least

CameronB9 replied:
How might a get_logger function work? I did consider a wrapper around the logger that automatically passes things like user ID by default. I thought it would add extra complexity due to the different logs passing different values depending on the context.

JackMeadDev replied:
Ah this is the question - many different options!

It might be that we just want a function that returns a logger, which uses the app logger if available, or falls back to something else for e.g. unit tests where it isn't:
```python
def get_custom_logger():
    if current_app:
        return current_app.logger
    else:
        return getLogger()
```

But if we want it to auto-populate some fields, we could take that further by creating a wrapper class, and adding our own `info` methods etc. that in turn call the underlying logger/info method, adding in their context

In thinking about this, I did realise I may be re-building something that already exists, and it looks like maybe [LoggerAdapters](https://docs.python.org/2/howto/logging-cookbook.html#adding-contextual-information-to-your-logging-output) are the textbook answer and basically help us achieve that without re-writing the info/warn functions

JackMeadDev commented on todo_app/app.py:
```diff
@@ -147,34 +193,61 @@ def add_todo():
+                "action": LogAction.add_todo,
+                "user_id": user.id,
+                "user_role": user.role,
+                "todo_id": todo_id, 
```
Nice, I can see you've carefully chosen to use id over the item description which is much safer from a sensitive-data perspective

JackMeadDev commented on todo_app/user.py:
```diff
@@ -39,7 +50,15 @@ def decorator(*args, **kwargs):
                         return fn(*args, **kwargs)
-                
+
+                app.logger.info({
```
I notice that this is a WARNING in your screenshot, is there any particular reason it was changed to info?

CameronB9 replied:
I did initially set it as warning but I wasn't really sure it should be warning? It could signal that someone is trying things they shouldn't so maybe should be a warning?

JackMeadDev replied:
Aha, this is one of those mean questions where I ask it and don't really care what the answer is - just that you've thought about it! 😄 
Argument for `info`: 
* we _might_ expect users to try things they're not supposed to do, that's a normal part of the app functionality, and therefore it's not a concern
* Anything else?
Argument for `warning`:
* Users shouldn't see UI options to do things they're not permitted to. Seeing this warning might therefore hint that our UI isn't correctly hiding something it should be, even if the backend is then preventing anything bad from actually happening
* Seeing lots of warnings might suggest something unusual or concerning, like an attempted attack

JackMeadDev commented on docker-compose.dev.yml:
```diff
@@ -47,4 +49,11 @@ services:
\ No newline at end of file
+      - ./dev_db:/data/db
+
+  prometheus:
```
Unrelated here, but I notice that when I ran this file I couldn't access the localhost address because the Dockerfile doesn't specify `--host=0.0.0.0` for the dev container

JackMeadDev reviewed (approved):
Nice work, you've got thorough and sensible logging added, and well done yet again on hitting the stretch goals! 😄 I've left some optional thoughts below, but very happy for you to merge when you're ready

JackMeadDev reviewed (commented):
(Responding to comments)
Oh, and bonus points if you can offer me any explanation for why your `json` logs show up in Loggly for me when I run the app through Docker, but not when I run it directly on my local machine! I'm 96% certain that I'm using the same environment variables and source code

Commit:
Updated dockerfile to include --host flag for development.
Add get logger function instead of passing app variable into user class method

CameronB9 commented:
> (Responding to comments) Oh, and bonus points if you can offer me any explanation for why your `json` logs show up in Loggly for me when I run the app through Docker, but not when I run it directly on my local machine! I'm 96% certain that I'm using the same environment variables and source code

I can't seem to run it locally. I have updated the env variables for the .env file to match the docker ones. I'm getting an DB timeout error, the address shown is the previous one I had in that file (the db which was created before terraform). It seems to be caching the .env file? I remember having this issue before when I was trying to run it locally to debug. I can't remember how I fixed it. 

I have had a quick look through the code and can't see anything that would cause it to not to log to loggly?

JackMeadDev commented:
On caching .env, I tend to blame VSCode first - so you could try opening a PowerShell/GitBash terminal and running the app through poetry there and seeing if that resolves the issue. I haven't found a satisfying solution to stop VSCode, but restarting it/the machine _may_ work, or make sure to clear out any potentially Python-cached files, such as `__pycache__` folders or similar
